File: go/src/kanzi/test/TestHash.go
/*
Copyright 2011-2013 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"flag"
	"fmt"
	"kanzi/util"
	"os"
	"time"
)

func main() {
	var filename = flag.String("input", "c:\\temp\\rt.jar", "name of the input file")
	flag.Parse()
	iter := 500
	fmt.Printf("Processing %v\n", *filename)
	fmt.Printf("%v iterations\n", iter)

	{
		fmt.Printf("XXHash speed test\n")
		file, err := os.Open(*filename)

		if err != nil {
			fmt.Printf("Cannot open %s", *filename)

			return
		}

		defer file.Close()
		buffer := make([]byte, 16384)
		hash, err := util.NewXXHash(uint32(0))

		if err != nil {
			fmt.Printf("Failed to create hash: %v\n", err)
			return
		}

		length, err := file.Read(buffer)
		size := int64(0)
		res := uint32(0)
		sum := int64(0)

		for length > 0 {
			if err != nil {
				fmt.Printf("Failed to read the next chunk of input file '%v': %v\n", *filename, err)
				return
			}

			before := time.Now()

			for i := 0; i < iter; i++ {
				hash.SetSeed(0)
				res += hash.Hash(buffer[0:length])
			}

			after := time.Now()
			sum += after.Sub(before).Nanoseconds()
			size += int64(length * iter)
			length, err = file.Read(buffer)
		}

		sum /= 1000000 // convert to ms
		fmt.Printf("XXHash res=%x\n", res)
		fmt.Printf("Elapsed [ms]: %v\n", sum)
		fmt.Printf("Throughput [MB/s]: %v\n", (size/1024*1000/1024)/sum)
	}

	fmt.Printf("\n")

	{
		fmt.Printf("MurmurHash3 speed test\n")
		file, err := os.Open(*filename)

		if err != nil {
			fmt.Printf("Cannot open %s", *filename)

			return
		}

		defer file.Close()
		buffer := make([]byte, 16384)
		hash, err := util.NewMurMurHash3(uint32(0))

		if err != nil {
			fmt.Printf("Failed to create hash: %v\n", err)
			return
		}

		length, err := file.Read(buffer)
		size := int64(0)
		res := uint32(0)
		sum := int64(0)

		for length > 0 {
			if err != nil {
				fmt.Printf("Failed to read the next chunk of input file '%v': %v\n", *filename, err)
				return
			}

			before := time.Now()

			for i := 0; i < iter; i++ {
				hash.SetSeed(0)
				res += hash.Hash(buffer[0:length])
			}

			after := time.Now()
			sum += after.Sub(before).Nanoseconds()
			size += int64(length * iter)
			length, err = file.Read(buffer)
		}

		sum /= 1000000 // convert to ms
		fmt.Printf("MurmurHash3 res=%x\n", res)
		fmt.Printf("Elapsed [ms]: %v\n", sum)
		fmt.Printf("Throughput [MB/s]: %v\n", (size/1024*1000/1024)/sum)
	}

	fmt.Printf("\n")

	{
		fmt.Printf("SipHash_2_4 speed test\n")
		file, err := os.Open(*filename)

		if err != nil {
			fmt.Printf("Cannot open %s", *filename)

			return
		}

		defer file.Close()
		buffer := make([]byte, 16384)
		hash, err := util.NewSipHash()

		if err != nil {
			fmt.Printf("Failed to create hash: %v\n", err)
			return
		}

		length, err := file.Read(buffer)
		size := int64(0)
		res := uint64(0)
		sum := int64(0)

		for length > 0 {
			if err != nil {
				fmt.Printf("Failed to read the next chunk of input file '%v': %v\n", *filename, err)
				return
			}

			before := time.Now()

			for i := 0; i < iter; i++ {
				hash.SetSeedFromLongs(0, 0)
				res += hash.Hash(buffer[0:length])
			}

			after := time.Now()
			sum += after.Sub(before).Nanoseconds()
			size += int64(length * iter)
			length, err = file.Read(buffer)
		}

		sum /= 1000000 // convert to ms
		fmt.Printf("SipHash_2_4 res=%x\n", res)
		fmt.Printf("Elapsed [ms]: %v\n", sum)
		fmt.Printf("Throughput [MB/s]: %v\n", (size/1024*1000/1024)/sum)
	}
}

File: java/src/kanzi/test/TestHash.java
/*
Copyright 2011-2013 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package kanzi.test;

import java.io.File;
import java.io.FileInputStream;
import kanzi.util.MurMurHash3;
import kanzi.util.XXHash;


public class TestHash
{
  public static void main(String[] args)
  {    
      try
      {
         String fileName = (args.length > 0) ? args[0] : "c:\\temp\\rt.jar";
         int iter = 500;
         System.out.println("Processing "+fileName);
         System.out.println(iter+" iterations");

         {
            System.out.println("XXHash speed test");
            File input = new File(fileName);
            FileInputStream fis = new FileInputStream(input);
            byte[] array = new byte[16384];
            XXHash hash = new XXHash(0);
            int len;
            long size = 0;
            int res = 0;
            long before = System.nanoTime();

            while ((len = fis.read(array, 0, array.length)) > 0)
            { 
               for (int i=0; i<iter; i++)
                  res += hash.hash(array, 0, len);
       
               size += (len * iter);
            }
           
            long after = System.nanoTime();
            fis.close();
            System.out.println("XXHash res="+Integer.toHexString(res));
            System.out.println("Elapsed [ms]: " +(after-before)/1000000L);
            System.out.println("Throughput [MB/s]: " +(size/1024*1000/1024)/((after-before)/1000000L));
         }
         
         System.out.println();
         
         {
            System.out.println("MurmurHash3 speed test");
            File input = new File(fileName);
            FileInputStream fis = new FileInputStream(input);
            byte[] array = new byte[16384];
            MurMurHash3 hash = new MurMurHash3(0);
            int len;
            long size = 0;
            int res = 0;
            long before = System.nanoTime();

            while ((len = fis.read(array, 0, array.length)) > 0)
            { 
               for (int i=0; i<iter; i++)
                  res += hash.hash(array, 0, len);     
               
               size += (len * iter);
            }            
            
            long after = System.nanoTime();
            fis.close();
            System.out.println("MurmurHash res="+Integer.toHexString(res));
            System.out.println("Elapsed [ms]: " +(after-before)/1000000L);
            System.out.println("Throughput [MB/s]: " +(size/1024*1000/1024)/((after-before)/1000000L));
         }
      }
      catch (Exception e)
      {
         e.printStackTrace();
      }
  }  
}